File: slurm2json/slurm2json.pl
#!/usr/bin/perl -wT
# vim: ts=4 sw=4 et
#
# slurm2json: Slurm config, QOS, and account converter.
#
# Written By:
# - A. Karl Kornel <akkornel@stanford.edu>
#
# Last Updated: 2017-05-15
#
# Copyright © 2017 the Board of Trustees of the Leland Stanford Junior University.
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

use strict;
use warnings;

use IPC::Open2;
use JSON;
use List::Util 1.45; # 1.45+ needed for uniq()
use Text::CSV;

my $DEBUG = 0;
my $SACCTMGR_PATH = '/usr/bin/sacctmgr';

# Wipe out our environment, for taint-safety.
%ENV = ();

# This is our cluster information.
my $cluster_name;
my @features;
my %gres;
my %partitions;
my %qos;
my (@access_all, %access_by_account, %access_by_group);

# Begin processing slurm.conf on standard input
while (1) {
    my $line = <STDIN>;
    if (!defined($line)) {
        print STDERR "No more lines to parse\n" if $DEBUG;
        last;
    }
    chomp $line;

    # Filter out stuff to ignore
    if ((length($line) == 0)
        or ($line =~ m|^#|)
        or ($line =~ m|^\s+$|)) {
        print STDERR "Skipping line $line\n" if $DEBUG;
        next;
    }

    print STDERR "Looking at $line\n" if $DEBUG;

    # Catch multi-line lines (that is, those ending in \)
    # A few cases have whitespace and/or comments after the \, so we must catch that.
    while ($line =~ m|\\\s*$|) {
        my $extra_line = <STDIN>;
        # Remove whitespace from start/end of line, and comments from end of line.
        # But, keep the ending backslash.
        $extra_line =~ s|^\s+||;
        $extra_line =~ s|\s+(#.*)?$||;
        $line =~ s|\\\s*$|$extra_line|;
        print STDERR "Appended line $extra_line\n" if $DEBUG;
    }

    # Figure out what type of line we have.
    my $line_type;
    if ($line =~ m|^(\w+)=(.+)$|) {
        $line_type = lc($1);
    } else {
        print "Could not parse the line $line\n";
        exit 1;
    }
    print STDERR "This line is a $line_type\n" if $DEBUG;

    # We only care about some things
   
    # We want to capture the ClusterName
    if ($line_type eq 'clustername') {
        $line =~ m|^ClusterName=(.+)$|i;
        $cluster_name = $1;
    }

    # For NodeName, we want the feature list, which we add to the global list.
    # (We'll handle deduplication later)
    # We also want to get Gres stuff
    elsif ($line_type eq 'nodename') {
        # Pull out the Feature list
        if ($line =~ m|Feature="(.+)"|i) {
            push @features, (split /,/, $1);
            print STDERR "Found features!\n" if $DEBUG;
        } else {
            print STDERR "No features found\n" if $DEBUG;
        }

        # Pull out the Gres list
        if ($line =~ m|Gres=([a-z0-9:,]+)|i) {
            my $gres_string = $1;
            print STDERR "Found Gres line $gres_string\n" if $DEBUG;
            foreach my $gres_entry (split(/,/, $gres_string)) {
                # Parse out the name, and optional type.
                my @gres_params = split(/:/, $gres_entry);
                my $gres_name = shift @gres_params;
                print STDERR "Gres name is $gres_name\n" if $DEBUG;

                # Make sure the top-level name exists, then add the option
                if (!exists($gres{$gres_name})) {
                    $gres{$gres_name} = [];
                }

                # If a Gres type is defined, add it to the list.
                # Again, we don't de-duplicate right now.
                if (scalar(@gres_params) > 0) {
                    push @{$gres{$gres_name}}, shift @gres_params;
                    print STDERR "Gres type is $gres_params[0]\n" if $DEBUG;
                }
            }
        } else {
            print STDERR "No Gres found\n" if $DEBUG;
        }
    }

    # For PartitionName, we want alot of stuff!
    elsif ($line_type eq 'partitionname') {
        my ($name, $options);
        my $hidden;
        my (%default, %max, %min, @qos);
        my (@allowed_accounts, @allowed_groups);

        # Get the partition name and the rest of the line
        if ($line =~ m|^PartitionName=(\w+)\s+(.+)$|i) {
            ($name, $options) = ($1, $2);
            print STDERR "Name is $name\nOptions are $options\n" if $DEBUG;
        } else {
            print "Could not extract name from $line\n";
            exit 1;
        }

        # Now, let's go through the rest of the options
        foreach my $option (split(/\s+/, $options)) {
            my ($option_name, $option_value) = split(/=/, $option);
            $option_name = lc($option_name);

            # We're going to look for options in this order:
            # Hidden, AllowQOS, Defaults (plus QOS), Max, Min, and Allow/Deny.
           
            # Catch Hidden
            if ($option_name eq 'hidden') {
                print STDERR "Found Hidden\n";
                if ($option_value eq 'YES') {
                    $hidden = JSON::true;
                }
                elsif ($option_value eq 'NO') {
                    $hidden = JSON::false;
                }
                else {
                    print "Invalid Hidden value \"$hidden\" for partition $name\n";
                    exit 1;
                }
            }

            # Build AllowQOS list
            elsif ($option_name eq 'allowqos') {
                @qos = split(/,/, $option_value);
            }

            # Defaults
            elsif ($option_name eq 'defmempercpu') {
                $default{'mem-per-cpu'} = $option_value;
            }
            elsif ($option_name eq 'defmempernode') {
                $default{'mem-per-node'} = $option_value;
            }
            elsif ($option_name eq 'defaulttime') {
                $default{'time'} = $option_value;
            }
            elsif ($option_name eq 'qos') {
                $default{'qos'} = $option_value;
            }

            # Maxima
            elsif ($option_name eq 'maxcpuspernode') {
                $max{'cpu-per-node'} = $option_value;
            }
            elsif ($option_name eq 'maxmempercpu') {
                $max{'mem-per-cpu'} = $option_value;
            }
            elsif ($option_name eq 'maxmempernode') {
                $max{'mem-per-node'} = $option_value;
            }
            elsif ($option_name eq 'maxtime') {
                $max{'time'} = $option_value;
            }
            elsif ($option_name eq 'maxnodes') {
                $max{'nodes'} = $option_value;
            }

            # Minima
            elsif ($option_name eq 'minnodes') {
                $min{'nodes'} = $option_value;
            }

            # Allowed entities
            elsif ($option_name eq 'allowaccounts') {
                @allowed_accounts = split(/,/, $option_value);
            }
            elsif ($option_name eq 'allowgroups') {
                @allowed_groups = split(/,/, $option_value);
            }

            # Denied entities
            # TODO: Add support for DenyAccounts
        }

        # Skip the DEFAULT partition
        if ($name eq 'DEFAULT') {
            print STDERR "Skipping DEFAULT partition\n" if $DEBUG;
            next;
        }

        # Build our partition entry
        my %partition_entry;
        if (defined($hidden)) {
            $partition_entry{'hidden'} = $hidden;
        }
        if (scalar(@qos) != 0) {
            $partition_entry{'qos'} = \@qos;
        }
        if (scalar(keys(%default)) != 0) {
            $partition_entry{'default'} = \%default;
        }
        if (scalar(keys(%max)) != 0) {
            $partition_entry{'max'} = \%max;
        }
        if (scalar(keys(%min)) != 0) {
            $partition_entry{'min'} = \%min;
        }

        # Add our partition to the list!
        $partitions{$name} = \%partition_entry;

        # Now we need to work out who can access this partition!
        
        # If there's nobody in the allowed list, then everyone can access!
        if (!scalar(@allowed_accounts) and !scalar(@allowed_groups)) {
            push @access_all, $name;
        }

        # Otherwise, add this partition to the access lists for accounts & groups.
        if (scalar(@allowed_accounts) > 0) {
            foreach my $account (@allowed_accounts) {
                if (!exists($access_by_account{$account})) {
                    $access_by_account{$account} = [$name];
                } else {
                    push @{$access_by_account{$account}}, $name;
                }
            }
        }
        if (scalar(@allowed_groups) > 0) {
            foreach my $group (@allowed_groups) {
                if (!exists($access_by_group{$group})) {
                    $access_by_group{$group} = [$name];
                } else {
                    push @{$access_by_group{$group}}, $name;
                }
            }
        }

    } # Done with PartitionName
} # Done processing slurm.conf

# Start running external commands!
my ($sacctmgr_pid, $sacctmgr_stdin, $sacctmgr_stdout);

# Pull QoS information from SLURM.

# To help, create an anonymous function that parses a TRES string.
my $tres_parse = sub {
    my ($tres_text) = (@_);
    my %output;

    foreach my $tres_item (split(/,/, $tres_text)) {
        # Each item is name=value
        my @tres_components = split(/=/, $tres_item);

        # For most items, processing is simple!
        if ($tres_components[0] !~ m|^gres/|) {
            $output{$tres_components[0]} = $tres_components[1];
        }

        # For gres components, populate a separate structure.
        else {
            if (!exists($output{'gres'})) {
                $output{'gres'} = {};
            }

            # First, split out the gres string
            $tres_components[0] =~ m|^gres/(.+)$|;
            my $gres_item = $1;

            # For now, put that entire item in.
            $output{'gres'}->{$gres_item} = $tres_components[1];
            # TODO: Properly split out the gres string.
        }
    }

    # Return our hashref!
    return \%output;
};

# Run sacctmgr to get the list of QOSes, in machine-readable form, with only what we want.
eval {
    my @cmdline = ($SACCTMGR_PATH, qw(-p list qos),
        'Format=Name,MaxTRESPerJob,MaxTRESPerUser,MaxWall,MinTRES',
    );
    print STDERR 'Running command: ', join(' ', @cmdline), "\n" if $DEBUG;
    $sacctmgr_pid = open2($sacctmgr_stdout, $sacctmgr_stdin, @cmdline);
    close($sacctmgr_stdin);
    print STDERR "PID is $sacctmgr_pid\n" if $DEBUG;
};
if ($@) {
    print STDERR "sacctmgr error: $@\n" if $DEBUG;
    print "WARNING!  Unable to run sacctmgr to get QOS info.\n";
    print "JSON will be incomplete.\n";
} else {
    # We'll use Text::CSV to handle parsing of sacctmgr's output.
    my $csv = Text::CSV->new({
        'sep_char'       => '|',
        'blank_is_undef' => 1,
        'empty_is_undef' => 1,
    });

    # Pull the first row for column names.
    $csv->column_names($csv->getline($sacctmgr_stdout));

    # Loop through each QoS line.
    while (my $qos_entry = $csv->getline_hr($sacctmgr_stdout)) {
        # Now we need to look at individual columns
        
        # First, get the name and create our entry.
        my $qos_name = $qos_entry->{'Name'};
        my (%qos_min, %qos_max);
        print STDERR "Examining QOS entry $qos_name\n" if $DEBUG;

        # MaxWall is easy to check for.
        if (defined($qos_entry->{'MaxWall'})) {
            $qos_max{'wall'} = $qos_entry->{'MaxWall'};
        }

        # Parse our minimums.
        if (defined($qos_entry->{'MinTRES'})) {
            %qos_min = %{$tres_parse->($qos_entry->{'MinTRES'})};
        }

        # Parse our per-user node maxima.
        if (defined($qos_entry->{'MaxTRESPU'})) {
            %qos_max = %{$tres_parse->($qos_entry->{'MaxTRESPU'})};
        }

        # For per-host maxima, if there's a conflict.
        # We assume that user-level overrides host-level TRES.
        # TODO: Check if this logic is actually correct.
        if (defined($qos_entry->{'MaxTRES'})) {
            my %qos_max_host = %{$tres_parse->($qos_entry->{'MaxTRESPU'})};

            foreach my $conflicting_qos (keys(%qos_max_host)) {
                if (!exists($qos_max{$conflicting_qos})) {
                    $qos_max{$conflicting_qos} = $qos_max_host{$conflicting_qos};
                }
            }
        }

        # Assemble our QOS components into the global hash.
        $qos{$qos_name} = {};
        if (scalar(keys(%qos_min)) > 0) {
            $qos{$qos_name}->{'min'} = \%qos_min;
        }
        if (scalar(keys(%qos_max)) > 0) {
            $qos{$qos_name}->{'max'} = \%qos_max;
        }
    } # Done looping through sacctmgr's QOS output.
} # Done with sacctmgr post-eval code.

# Clean up sacctmgr's child process.
waitpid($sacctmgr_pid, 0);


# Start building our JSON.


# Catch if we didn't get a cluster name
if (!defined $cluster_name) {
    print "No ClusterName line found.  Did you provide a config?\n";
    exit 1;
}

# Make our cluster hash, starting with partitions.
my %cluster = (
    partitions => \%partitions,
);

# If we have a populated QOS hash, add them.
if (scalar(keys(%qos)) > 0) {
    $cluster{'qos'} = \%qos;
}

# If we have features, deduplicate and add them.
# NOTE: This doesn't sort the list!
if (scalar(@features) > 0) {
    $cluster{'features'} = [List::Util::uniq(@features)];
}

# If we have Gres, add them as well.
# If a key has any types, add them as well.  Else add an empty list.
# And again, NOTE, this doesn't sort!
if (scalar(keys(%gres)) > 0) {
    foreach my $gres_key (keys(%gres)) {
        my @gres_types = List::Util::uniq(@{$gres{$gres_key}});
        $gres{$gres_key} = \@gres_types;
    }
    $cluster{'gres'} = \%gres;
}

# If we have any users, then set up that hash key.
if ((scalar(@access_all) > 0) or
    (scalar(keys(%access_by_account)) > 0) or
    (scalar(keys(%access_by_group)))
) {
    $cluster{'users'} = {};
}

# Add the wildcard user
if (scalar(@access_all) > 0) {
    $cluster{'users'}->{'*'} = \@access_all;
}

# TODO: Handle %access_by_account and %access_by_group

# Output the JSON: We encode as UTF-8, and we pretty-print
print JSON->new->utf8(1)->pretty(1)->encode({ $cluster_name => \%cluster });

print STDERR "All done!\n" if $DEBUG;
exit 0;

File: helpers/slurm2json.py
#!/usr/bin/env python

'''
slurm2json.py: convert a slurm.conf to a machines.json input file

The MIT License (MIT)

Copyright (c) 2017 Vanessa Sochat

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

'''

import argparse
import json
import sys
import os
import re


def get_parser():
    parser = argparse.ArgumentParser(
    description="convert slurm.conf to machines.json")

    parser.add_argument("--input", dest='input', 
                        help='''path to one or more slurm config files, separated by commas. 
                                Default is one slurm.conf in present working directory.''', 
                        type=str, default='slurm.conf')

    parser.add_argument("--update", dest='update', 
                        help='''Update an already existing machines.json (or other)''', 
                        default=False, action='store_true')


    parser.add_argument("--disclude-part", dest='disclude_part', 
                        help='''Partitions to disclude, separated by commas''', 
                        type=str, default=None)


    parser.add_argument("--print", dest='print', 
                        help="print to screen instead of saving to machines.json", 
                        default=False, action='store_true')


    parser.add_argument("--quiet", dest='quiet', 
                        help='''Suppress all output (other than print)''', 
                        default=False, action='store_true')


    parser.add_argument("--force", dest='force', 
                        help="Force overwrite of the output file, if it exists.", 
                        default=False, action='store_true')

    # Two images, for similarity function
    parser.add_argument("--outfile", dest='outfile', 
                        help="output json file. Default is machines.json", 
                        type=str, default='machines.json')

    return parser




def main():

    parser = get_parser()

    try:
        args = parser.parse_args()
    except:
        sys.exit(0)

    if os.path.exists(args.outfile) and args.force is False and args.update is False and args.print is False:
        message("%s already exists! Use --force to force overwrite." %args.outfile)
        sys.exit(1)

    print("Parsing %s, please wait!" %args.input)

    # If the user wants an update, the output file must exist
    if args.update is True:
        if not os.path.exists(args.outfile):
            message("Cannot find %s to update. Did you specify the right path?" %args.outfile)
            sys.exit(1)
        message("Found %s to update." %args.outfile,args.quiet)
        machines = read_json(args.outfile)
    else:
        machines = dict()

    # Does the user want to disclude partitions?
    disclude_part = None
    if args.disclude_part is not None:
        disclude_part = args.disclude_part.split(',')
        message("%s will not be included." %', '.join(disclude_part),args.quiet)
    else:
        message("All partitions will be included.",args.quiet)

    input_files = args.input.split(',')
    for input_file in input_files: 
        if not os.path.exists(input_file):
            message("Cannot find %s. Did you specify the right path?" %input_file)
            sys.exit(1)
        cluster = parse_config(config_file=input_file)
        cluster_names = ",".join(list(cluster.keys()))
        message('Adding cluster %s' %(cluster_names),args.quiet)
        if disclude_part is not None:
            cluster = disclude_partitions(cluster,disclude_part)  
        machines.update(cluster)

    cluster_names = ",".join(list(machines.keys()))
    message('Compiling clusters %s' %(cluster_names),args.quiet)
    if args.print is True:
        message(json.dumps(machines, indent=4, sort_keys=True))
    else:
        write_json(machines,args.outfile)


########################################################################
# Utils
########################################################################

def message(text,quiet=False):
    if not quiet:
        print(text)

def unpack_data(data):
    config = data['config']
    nodes = data['nodes']
    partitions = data['partitions']
    return config,nodes,partitions


def pack_data(config,nodes,partitions):
    return {'config':config,
            'nodes':nodes,
            'partitions':partitions}


def combine_clusters(clusters):
    machines = dict()
    for cluster in clusters:
        for cluster_name,metadata in cluster:
            machines[cluster_name] = metadata
    return machines


def read_file(file_name, clean=True, join=False):
    '''read in a file, with optional "clean up" to remove
    comments (lines starting with #) and empty lines)
    '''
    with open(file_name,'r') as filey:
        content = filey.readlines()
    if clean is True:
        content = [c.strip('\n') 
                   for c in content 
                   if not c.startswith('#') 
                   and len(c.strip('\n')) > 0]

    if join:
        content = ''.join(content)    
    return content


def write_json(json_obj,filename,mode="w"):
    with open(filename,mode) as filey:
        filey.writelines(json.dumps(json_obj, indent=4, separators=(',', ': ')))
    return filename


def read_json(filename,mode='r'):
    with open(filename,mode) as filey:
        data = json.load(filey)
    return data


def remove_comments(line):
    return line.rsplit('#',1)[0].strip()


def parse_line_multi(line,keepers=None):
    '''parse_line_multiple will return a dictionary with keys and
    values for a line from a slurm conf, with variables expected to be separated
    by spaces. If keepers is not defined, all is kept'''
    parsed = dict()
    lines = line.strip().split('#')[0].split(' ')
    for line in lines:
        if len(line) > 0:
            params = line.split('=')
            key = params[0]
            key = "%s%s" %(key[0].capitalize(),key[1:])
            value = params[-1].strip()
            if keepers is not None:
                if key in keepers:
                    parsed[key] = value
            else:
                if key is not "\\":
                    parsed[key] = value
    return parsed



########################################################################
# Nodes
########################################################################

def get_node_variables():
    return ["RealMemory",
            "Gres",
            "Weight",
            "Feature",
            "Default"]



def break_range_expressions(node_name):
    parts = list(node_name)
    current = ''
    finished = []
    opened = False
    for c in range(len(parts)):
        part = parts[c]
        if part == '{':
            if len(current) > 0: 
                finished.append(current)       
            opened = True
            current='{'
        elif part == '}':
            if len(current) > 0:
                finished.append("%s}" %current)
                current=''       
            opened = False
        else:
            current = "%s%s" %(current,part)
              
    if opened:
        current = "%s}" %(current)
    if current not in finished and len(current)>0:
        finished.append(current)
    return finished
        

def parse_single_node(node_name):
    '''this function will parse a single string to describe a group of 
    nodes, eg gpu-27-{21,35}'''
    parts = break_range_expressions(node_name)
    options = []
    for part in parts:
        node_options = []
        if not re.search("^{|}$",part):  
            options.append([part])
        else:
            node_ranges = re.findall("[0-9]+-[0-9]+",part)
            node_lists = re.findall("[0-9]+,[0-9]+",part)
            for node_range in node_ranges:
                start,end = [int(x) for x in node_range.split('-')]
                node_options += [int(x) for x in range(start,end+1)]
            for node_list in node_lists:
                node_options += [int(x) for x in node_list.split(',')]
            options.append(node_options)
    final_options = options.pop(0)
    while len(options) > 0:
        option_set = options.pop(0)
        new_options = []
        for final_option in final_options:
            for option in option_set:
                new_options.append("%s%s" %(final_option,option))
        final_options = new_options
    return final_options


def parse_node_names(line):
    '''parse_node_names will take a whole list of nodes (multiple with ranges and
    lists in brackets) and return a list of unique, complete names.'''
    new_nodes = []
    #nodelist = re.sub("\\\\| ","",line).split('=')[-1]
    nodelist = re.sub("\\\\| ","",line)
    nodelist = nodelist.replace('[','{').replace(']','}')
    nodelist = re.split(',\s*(?![^{}]*\})', nodelist)
    for node_name in nodelist:
        contenders = [x for x in parse_single_node(node_name) if x not in new_nodes]
        new_nodes = new_nodes + contenders
    return new_nodes

def keep_going(name):
    '''A function to filter a string to determine if
    the calling function should continue. Returns False
    if the string being checked contains any flag variables.
    '''
    skip_these = ['DEFAULT','test']
    go_on = True
    for skip in skip_these:
        if name.startswith(skip):
            go_on = False
    return go_on


def parse_node_block(data):
    '''line should be the first line popped that has 'NodeName'
    and config is the entire config following that. The new node
    entry is added to the global nodes.
    '''
    config,nodes,partitions = unpack_data(data)

    line = parse_line_multi(config.pop(0))

    if "NodeName" not in line:
        return pack_data(config,nodes,partitions)

    node = line['NodeName']
    if not keep_going(node):
        return pack_data(config,nodes,partitions)
    
    # Get all variables for node group
    keepers = get_node_variables()
    node_names = parse_node_names(node)
    del line['NodeName']
    node_settings = line
    done = False

    while not done:
        line = remove_comments(config.pop(0))
        if not line.endswith('\\'):
            done = True
        updates = parse_line_multi(line,keepers)
        node_settings.update(updates)

    for node in node_names:
        if node not in nodes:
            nodes[node] = node_settings
            nodes[node]['partitions'] = [] 
        else:
            nodes[node].update(node_settings)    
    return pack_data(config,nodes,partitions)


########################################################################
# Features and Defaults
########################################################################

def parse_features(data):
    config,nodes,partitions = unpack_data(data)
    features = dict()
    for node_name,attributes in nodes.items():
        if 'Feature' in attributes:
            feature_list = attributes['Feature'].strip('"').split(',')
            for partition in attributes['partitions']:
                if partition not in features:
                    features[partition] = feature_list
                else:
                    new_features = [x for x in feature_list 
                                    if x not in features[partition]]
                    features[partition] = features[partition] + new_features
    return features


def find_defaults(data):
    defaults = dict()
    for key,datum in data.items():
        if datum:
            defaults[key] = []
            for name,attributes in datum.items():
                if "Default" in attributes:
                    defaults[key].append(name)
    return defaults

########################################################################
# Partitions
########################################################################

def get_partition_variables():
    return ["DefaultTime",
            "Default",
            "DefMemPerCPU",
            "MaxMemPerCPU",
            "AllowQos",
            "Nodes"]


def disclude_partitions(cluster,disclude_parts):
    for cluster_name, attributes in cluster.items():
        if "partitions" in attributes:
            for disclude_part in disclude_parts:
                if disclude_part in attributes['partitions']:
                    del cluster[cluster_name]['partitions'][disclude_part]
        if "nodes" in attributes:
            for node_name, node_attributes in attributes['nodes'].items():
                if "partitions" in node_attributes:    
                    update = [x for x in cluster[cluster_name]['nodes'][node_name]['partitions']
                              if x not in disclude_parts]
                    cluster[cluster_name]['nodes'][node_name]['partitions'] = update
        if "features" in attributes:
            for disclude_part in disclude_parts:
                if disclude_part in attributes['features']:
                    del cluster[cluster_name]['features'][disclude_part]
    return cluster


def parse_partition_block(data):
    '''line should be the first line popped that has 'PartitionName'
    and config is the entire config following that. The new partition
    entry is added to the global partitions. The next line (non partition)
    is returned.
    '''
    config,nodes,partitions = unpack_data(data)

    line = parse_line_multi(config.pop(0))
 
    if "PartitionName" not in line:
        return pack_data(config,nodes,partitions)

    partition_name = line['PartitionName']
    keepers = get_partition_variables()

    if not keep_going(partition_name):
        return pack_data(config,nodes,partitions)

    # Get all variables for node group
    new_partition = line
    done = False
    while not done:
        line = remove_comments(config.pop(0))
        if not line.endswith('\\'):
            done = True
        updates = parse_line_multi(line,keepers)
        if "Nodes" in updates:
            parts = parse_node_names(updates['Nodes'])
            for node in parts:
                if node in nodes:
                    if partition_name not in nodes[node]['partitions']:
                        nodes[node]['partitions'].append(partition_name)
                else:
                    nodes[node] = {'partitions':[partition_name]} 
            # Note, we don't add an exaustive list of nodes to each partition
            # But if we needed to, could do that here.
            updates['maxNodes'] = len(parts)
            del updates['Nodes']
        new_partition.update(updates)    
    partitions[partition_name] = new_partition
    return pack_data(config,nodes,partitions)


########################################################################
# Main Parser
########################################################################

def parse_config(config_file):
    '''parse a config file to return a complete list of machines
    '''
    machines = dict()
    config = read_file(config_file)    
    data = {'partitions':{},
            'nodes': {},
            'config':config}

    while data['config']:
        line = data['config'][0]
        if line.startswith('ClusterName'):
            line = data['config'].pop(0)
            cluster = parse_line_multi(line)['ClusterName']
        elif line.startswith('PartitionName'):
            data = parse_partition_block(data)
        elif line.startswith('NodeName'):
            data = parse_node_block(data)        
        else:
            data['config'].pop(0)

    # Calculate features for each partition
    machines[cluster] = dict()
    machines[cluster]['features'] = parse_features(data)

    # Find Defaults
    machines[cluster]['defaults'] = find_defaults(data)

    del data['config']
    machines[cluster].update(data)



    return machines


if __name__ == '__main__':
    main()
